src/assets/JSAsset.js

```diff
@@ -117,16 +118,41 @@ class JSAsset extends Asset {
   }
 
   generate() {
-    // TODO: source maps
-    let code = this.isAstDirty
-      ? generate(this.ast).code
-      : this.outputCode || this.contents;
+    let code;
+    if (this.isAstDirty) {
+      let opts = {
+        sourceMaps: this.options.sourcemaps,
+        sourceFileName: this.options.sourcemaps ? this.relativename : undefined
+      };
+      let generated = generate(this.ast, opts, this.contents);
+      if (this.sourcemap && generated.map) {
```

Comment:
When/why does this case occur (there is already a source map)?

Reply:
This is when babel has to generate the code from ast after for example Typescript parsed the code. Typescript sets a sourcemap, although babel changes the sourcemap in these particular cases, and this function offsets it correctly again.
This happens in the typescript env test for example